### ISSUE TITLE
Using BF16 precision on cloud TPUs

### ISSUE THREAD

**Issue by deepakkumar2440** [pytorch/xla#6359]
## ❓ Questions and Help
Hi,
I am trying to benchmark/workload analysis on TPUs. When running Resneg 50, I get the following results:
Batch Size 8: Average Inference Time: 0.0227 seconds
Batch Size 16: Average Inference Time: 0.0269 seconds
Batch Size 32: Average Inference Time: 0.0351 seconds
Batch Size 64: Average Inference Time: 0.0512 seconds
Batch Size 128: Average Inference Time: 0.0749 seconds
Batch Size 256: Average Inference Time: 0.1327 seconds

Then I  set the env variable using %env XLA_USE_BF16=1. I ran the same expiriment and didnt see any difference. 
Batch Size 8: Average Inference Time: 0.0229 seconds
Batch Size 16: Average Inference Time: 0.0271 seconds
Batch Size 32: Average Inference Time: 0.0352 seconds
Batch Size 64: Average Inference Time: 0.0515 seconds
Batch Size 128: Average Inference Time: 0.0751 seconds
Batch Size 256: Average Inference Time: 0.1326 seconds

Is there anything that I am missing to run the calculations in BF16? 

My code for both scenarios:

import torchvision
import matplotlib.pyplot as plt
import numpy
import time
import torch
list_infer = []
list_avg = []
# Load your model and input data
model = torchvision.models.resnet50(pretrained=True)
device = xm.xla_device()
xla_model= model.to(device)
inference_time = 0
batch_sizes = [8, 16, 32, 64, 128, 256, 512]

for batch_size in batch_sizes:
    input_data = torch.randn(batch_size, 3, 224, 224).to(device)  # Adjust the shape as per your model's input
    model.eval()
    for x in range(0,3)  : 
        with torch.no_grad():
                output = xla_model(input_data)
    for x in range(0,300):
    # Put the model in evaluation mode (no gradient calculation)
        model.eval()
    
        # Measure inference time
        with torch.no_grad():
             start_time = time.time()
             output = xla_model(input_data)
             xm.mark_step() # execute the model
             xm.wait_device_ops()
             end_time = time.time()
    
        inference_time = inference_time + end_time - start_time
        infer_t =   (end_time - start_time)
        list_infer.append(infer_t) 
    avg = numpy.average(list_infer) 
    list_avg.append(avg)
    print(f"Batch Size {batch_size}: Average Inference Time: {avg:.4f} seconds")
    avg = 0
    list_infer = []
       #for that iteration
        
fig, ax = plt.subplots()

#fruits = ['', '', 'cherry', 'orange']
#counts = [40, 100, 30, 55]
#bar_labels = ['', 'blue', '_red', 'orange']
bar_colors = ['tab:red', 'tab:blue', 'tab:red', 'tab:orange']

ax.scatter(batch_sizes,  (list_avg))

ax.set_ylabel('higher_avg')
ax.set_xlabel('batchsize')
ax.set_title('comparision of inference time vs batchsize')
    
    
    #thrpt = 5 / inference_time
    #print("Throughput : " , thrpt)

**Comment by JackCaoG** [comment]
@alanwaketan can you take a look?

**Comment by alanwaketan** [comment]
Can you change the dtype in the modeling code to bf16 as well?
https://pytorch.org/docs/stable/generated/torch.set_default_dtype.html

**Comment by deepakkumar2440** [comment]
> Can you change the dtype in the modeling code to bf16 as well? https://pytorch.org/docs/stable/generated/torch.set_default_dtype.html

Hi, 
Thank you for your reply :).
I have done as you asked. here are my results. Do they make sense or am I still doing something wrong?
**FP32 (XLA_USE_BF16 not set and default dtype float32):**
Batch Size 2: Average Inference Time: 0.0174 seconds
Batch Size 8: Average Inference Time: 0.0225 seconds
Batch Size 16: Average Inference Time: 0.0273 seconds
Batch Size 32: Average Inference Time: 0.0375 seconds
Batch Size 64: Average Inference Time: 0.0571 seconds
Batch Size 128: Average Inference Time: 0.1012 seconds
Batch Size 256: Average Inference Time: 0.1781 seconds
Batch Size 512: Average Inference Time: 0.3322 seconds

**BF16 (XLA_USE_BF16  set and default dtype bfloat16):**
Batch Size 2: Average Inference Time: 0.0170 seconds
Batch Size 8: Average Inference Time: 0.0214 seconds
Batch Size 16: Average Inference Time: 0.0257 seconds
Batch Size 32: Average Inference Time: 0.0341 seconds
Batch Size 64: Average Inference Time: 0.0511 seconds
Batch Size 128: Average Inference Time: 0.0922 seconds
Batch Size 256: Average Inference Time: 0.1957 seconds
Batch Size 512: Average Inference Time: 0.3359 seconds

**Comment by alanwaketan** [comment]
The way you time your script is actually including the tracing time.

I will either use a profile to measure the execution time or put another start_time after the mark_step call and before the wait_device_ops call.
